src/ngx_http_lua_shdict.c

```diff
@@ -1305,6 +1305,461 @@ ngx_http_lua_ffi_shdict_udata_to_zone(void *zone_udata)
 }
 
 
+int
+ngx_http_lua_ffi_shdict_cas(ngx_shm_zone_t *zone, u_char *key,
+    size_t key_len, int old_value_type, u_char *old_str_value_buf,
+    size_t old_str_value_len, double old_num_value, int old_user_flags,
+    int value_type, u_char *str_value_buf, size_t str_value_len,
+    double num_value, int user_flags, int set_user_flags,
+    long exptime, int match_flags, int *match, char **errmsg, int *forcible)
+{
+    int                          i, n, match_value;
+    u_char                       c, oc, *p;
+    uint32_t                     hash;
+    ngx_int_t                    rc;
+    ngx_time_t                  *tp;
+    ngx_queue_t                 *queue, *q;
+    ngx_rbtree_node_t           *node;
+    ngx_http_lua_shdict_ctx_t   *ctx;
+    ngx_http_lua_shdict_node_t  *sd;
+
+    dd("exptime: %ld", exptime);
+
+    ctx = zone->data;
+
+    *match = 0;
+    *forcible = 0;
+    *errmsg = NULL;
+
+    hash = ngx_crc32_short(key, key_len);
+
+    switch (value_type) {
+
+    case SHDICT_TSTRING:
+        /* do nothing */
+        break;
+
+    case SHDICT_TNUMBER:
+        dd("num value: %lf", num_value);
+        str_value_buf = (u_char *) &num_value;
+        str_value_len = sizeof(double);
+        break;
+
+    case SHDICT_TBOOLEAN:
+        c = num_value ? 1 : 0;
+        str_value_buf = &c;
+        str_value_len = sizeof(u_char);
+        break;
+
+    case LUA_TNIL:
+        str_value_buf = NULL;
+        str_value_len = 0;
+        break;
+
+    default:
+        *errmsg = "unsupported value type";
+        return NGX_ERROR;
+    }
+
+    switch (old_value_type) {
+
+    case SHDICT_TSTRING:
+        /* do nothing */
+        match_value = 1;
+        break;
+
+    case SHDICT_TNUMBER:
+        dd("num value: %lf", old_num_value);
+        old_str_value_buf = (u_char *) &old_num_value;
+        old_str_value_len = sizeof(double);
+        match_value = 1;
+        break;
+
+    case SHDICT_TBOOLEAN:
+        oc = old_num_value ? 1 : 0;
+        old_str_value_buf = &oc;
+        old_str_value_len = sizeof(u_char);
+        match_value = 1;
+        break;
+
+    case LUA_TNIL:
+        match_value = 0;
+        break;
+
+    default:
+        *errmsg = "unsupported old_value type";
+        return NGX_ERROR;
+    }
+
+    ngx_shmtx_lock(&ctx->shpool->mutex);
+
+#if 1
+    ngx_http_lua_shdict_expire(ctx, 1);
+#endif
+
+    rc = ngx_http_lua_shdict_lookup(zone, hash, key, key_len, &sd);
+
+    dd("lookup returns %d", (int) rc);
+
+    if ((match_flags || match_value) &&
+        ((rc == NGX_DECLINED || rc == NGX_DONE) ||
+         (match_flags && old_user_flags != (int) sd->user_flags) ||
+         (match_value && (old_value_type != sd->value_type ||
+                old_str_value_len != sd->value_len ||
+                ngx_memcmp(old_str_value_buf, sd->data + key_len,
+                           (size_t) old_str_value_len)))))
```

Comment:
this is too huge, maybe we can split them into two `if` conditions?

Reply:
I could rewirte it to a lot more readable, but some code duplication.
```suggestion
    if (match_flags || match_value) {
        if (rc != NGX_OK) {
            ngx_shmtx_unlock(&ctx->shpool->mutex);
            *match = 1;
            return NGX_DECLINED;
        }

        if (match_flags && old_user_flags != (int) sd->user_flags) {
            ngx_shmtx_unlock(&ctx->shpool->mutex);
            *match = 1;
            return NGX_DECLINED;
        }

        if (match_value) {
            if (old_value_type != sd->value_type
                || old_str_value_len != sd->value_len
                || ngx_memcmp(old_str_value_buf, sd->data + key_len,
                              (size_t) old_str_value_len))
            {
                ngx_shmtx_unlock(&ctx->shpool->mutex);
                *match = 1;
                return NGX_DECLINED;
            }
        }
    }
```

Reply:
I think this may be better
```c
    if (match_flags
        && (rc != NGX_OK || old_user_flags != (int) sd->user_flags))
    {
        ngx_shmtx_unlock(&ctx->shpool->mutex);
        *match = 1;
        return NGX_DECLINED;
    }

    if (match_value
        && (rc != NGX_OK
            || old_value_type != sd->value_type
            || old_str_value_len != sd->value_len
            || ngx_memcmp(old_str_value_buf, sd->data + key_len,
                           (size_t) old_str_value_len)
        ))
    {
        ngx_shmtx_unlock(&ctx->shpool->mutex);
        *match = 1;
        return NGX_DECLINED;
    }
```